templates/Xamarin/Features/TelemetryHockeyApp.Prism/Param_ProjectName/Services/TrackingService.cs

```diff
@@ -0,0 +1,295 @@
+﻿using System;
+using System.Collections.Generic;
+using System.ComponentModel;
+using System.Diagnostics;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using Xamarin.Forms;
+
+namespace Param_ItemNamespace.Services
+{
+    public class TrackingService : ITrackingService
+    {
+        public TrackingLoggingLevel AppLoggingLevel { get; set; } = TrackingLoggingLevel.Information;
+        public TrackingLoggingLevel TrackEventLogLevel { get; set; } = TrackingLoggingLevel.Information;
+        public TrackingLoggingLevel TrackDependencyLogLevel { get; set; } = TrackingLoggingLevel.Information;
+        public TrackingLoggingLevel TrackPageViewLogLevel { get; set; } = TrackingLoggingLevel.Information;
+        public TrackingLoggingLevel TrackExceptionLogLevel { get; set; } = TrackingLoggingLevel.Error;
+        public TrackingLoggingLevel TrackMetricLogLevel { get; set; } = TrackingLoggingLevel.Warning;
+        public int AggregateMetricLoggingDelay { get; set; } = 60000;
+        protected readonly Dictionary<string, TrackingServiceMetric> TrackingServiceMetricCollection = new Dictionary<string, TrackingServiceMetric>();
+        public BackgroundWorker thread;
+        protected readonly Boolean stop = false;
+        protected readonly object _lockObject = new object();
+        protected ITrackingAPI TrackingAPI { get; private set; }
+
+        public TrackingService(ITrackingAPI trackingAPI)
+        {
+            TrackingAPI = trackingAPI;
+            thread = new BackgroundWorker();
+            thread.DoWork += async (o, e) =>
+            {
+                while (!stop)
+                {
+                    await Task.Delay(AggregateMetricLoggingDelay).ConfigureAwait(false);
+                    lock (_lockObject)
+                    {
+                        foreach (var item in TrackingServiceMetricCollection)
+                        {
+                            if (item.Value.Measurements.Count > 0 && item.Value.Start.Add(item.Value.MetricTimeSpan) <= DateTime.Now && AppLoggingLevel >= item.Value.MetricLoggingLevel)
+                            {
+                                Debug.WriteLine($"TrackingMetric sent - metric:{item.Key}, Number of Measurements:{ item.Value.Measurements.Count}");
+                                var metricMeasureTuple = Aggregrate(item.Key);
+                                Device.BeginInvokeOnMainThread(() => TrackingAPI.TrackEvent($"Metric:{item.Key}", metricMeasureTuple.Item1, metricMeasureTuple.Item2));
```

Comment:
Is there a requirement to track the event on the main thread?  If not, I'd keep it tracking on a separate thread if possible.

Reply:
Happens on only HockeyApp API’s for iOS, but I didn’t want to do this differently for Android/UWP (which work on the background thread and same with AppCenter API on all platforms).

Reply:
Fair enough.